dockerfile

```diff
@@ -0,0 +1,16 @@
+# base image
+FROM node:12.2.0-alpine
+
+# set working directory
+WORKDIR /app
+
+# add `/app/node_modules/.bin` to $PATH
+ENV PATH /app/node_modules/.bin:$PATH
+
+# install and cache app dependencies
+COPY package.json /app/package.json
```

Comment:
Waarom alleen de `package.json`?

Reply:
Toevoegen van de package.json (en vervolgens deze installeren via NPM install) zorgt er voor dat je altijd de juiste versies van de node_modules installeert (dus ook afgestemd op het operating system van de image). Daarnaast zorgt het ook voor caching van de node_modules via het layer-system van docker (als er niets wijzigt, gebruikt docker de bestaande layer ipv die telkens opnieuw te bouwen).
De code van de applicatie zelf wordt in een volgende stap toegevoegd om de zelfde reden. Zo ontstaan er 2 (opeenvolgende) layers waarbij alleen de buitenste layer (dus de applicatie-layer) wordt herbouwd als je iets in de code wijzigt en niet ook nog eens de node_modules moeten worden geïnstalleerd.